File: packages/scruple-host-sdk/scruple_host_sdk/queue.py
"""Offline retry queue for network calls that could not complete.

JSONL on disk, one line per queued request. If the host process crashes
between a failed submit() and the next successful drain, the queue
survives -- it is a file, not an in-memory list.

This is a near-direct port of the identical queue_store.py that existed,
byte-for-byte 93% identical, in all six forks -- tested in every one of
them and wired into the failure path in NONE of them (D-10, CANON_SKELETON
§5 property 3). The port itself is not the fix; http.submit() calling
enqueue() unconditionally on failure is the fix. See http.py.
"""

from __future__ import annotations

import json
import os
import time
import uuid
from typing import Any, Callable, Dict, List, Optional

BACKOFF_SCHEDULE = [5, 30, 120, 600, 1800]


class QueueStore:
    def __init__(self, path: str) -> None:
        self.path = path
        d = os.path.dirname(path)
        if d:
            os.makedirs(d, exist_ok=True)
        if not os.path.exists(path):
            with open(path, "w", encoding="utf-8"):
                pass

    def enqueue(
        self,
        *,
        kind: str,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        entry = {
            "id": uuid.uuid4().hex,
            "kind": kind,
            "method": method,
            "path": path,
            "body": body,
            "queued_at": time.time(),
            "attempts": 0,
            "last_attempt_at": None,
        }
        with open(self.path, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry) + "\n")
        return entry

    def load_all(self) -> List[Dict[str, Any]]:
        out: List[Dict[str, Any]] = []
        if not os.path.exists(self.path):
            return out
        with open(self.path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    out.append(json.loads(line))
                except json.JSONDecodeError:
                    continue
        return out

    def replace_all(self, entries: List[Dict[str, Any]]) -> None:
        tmp = self.path + ".tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            for e in entries:
                f.write(json.dumps(e) + "\n")
        os.replace(tmp, self.path)

    def count(self) -> int:
        return len(self.load_all())

    def due_now(self, now: Optional[float] = None) -> List[Dict[str, Any]]:
        n = now if now is not None else time.time()
        out: List[Dict[str, Any]] = []
        for e in self.load_all():
            if e.get("last_attempt_at") is None:
                out.append(e)
                continue
            idx = min(max(e["attempts"] - 1, 0), len(BACKOFF_SCHEDULE) - 1)
            wait = BACKOFF_SCHEDULE[idx]
            if n - e["last_attempt_at"] >= wait:
                out.append(e)
        return out

    def mark_attempted(self, entry: Dict[str, Any]) -> None:
        all_entries = self.load_all()
        for e in all_entries:
            if e["id"] == entry["id"]:
                e["attempts"] += 1
                e["last_attempt_at"] = time.time()
        self.replace_all(all_entries)

    def remove(self, entry: Dict[str, Any]) -> None:
        all_entries = self.load_all()
        kept = [e for e in all_entries if e["id"] != entry["id"]]
        self.replace_all(kept)

    def drain(
        self,
        submit_fn: Callable[[Dict[str, Any]], Any],
        *,
        now: Optional[float] = None,
    ) -> Dict[str, int]:
        """Retry every due entry through `submit_fn`, which must return
        something with a truthy/falsy `.ok`. Callers (Client.detach())
        MUST pass a submit_fn that calls http.submit() with
        queue_kind=None -- draining an already-queued entry must never
        re-enqueue a second copy of itself on repeated failure. Backoff
        state (mark_attempted) already tracks the retry schedule; a
        second enqueue would duplicate the entry rather than reschedule
        it.
        """
        succeeded = 0
        failed = 0
        for entry in self.due_now(now=now):
            self.mark_attempted(entry)
            result = submit_fn(entry)
            if getattr(result, "ok", False):
                self.remove(entry)
                succeeded += 1
            else:
                failed += 1
        return {"succeeded": succeeded, "failed": failed, "remaining": self.count()}

File: services/scruple-capture/src/queue.ts
// The durable offline queue (§5).
//
// WHY THIS IS A PORT AND NOT A NEW DESIGN. packages/scruple-host-sdk/queue.py
// already implements the durable half — JSONL on disk, one line per queued
// request, survives process death because it is a file and not an in-memory
// list, with BACKOFF_SCHEDULE = [5, 30, 120, 600, 1800]. Its own docstring
// records that the identical file was ported into all six SDK forks and wired
// into the failure path in NONE of them. The port is not the fix; calling
// enqueue() unconditionally on submission failure is the fix. That call is in
// submitter.ts.
//
// The component is Node, so it cannot import queue.py. The on-disk FORMAT and
// the schedule are kept identical to the Python file, field for field, so an
// operator can read either queue with either tool and so a future consolidation
// is a deletion rather than a migration. If you change a field name here,
// change it there.
//
// ONE FIELD THE PYTHON QUEUE DOES NOT HAVE: `counter`. It is not decoration.
// §5 says a queued event KEEPS ITS COUNTER — the counter was spent when the
// MAC was computed, and the MAC only verifies against the key at that counter.
// A queue that re-derived or re-numbered on drain would invalidate every
// entry it held.

import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

/** Identical to packages/scruple-host-sdk/scruple_host_sdk/queue.py. */
export const BACKOFF_SCHEDULE = [5, 30, 120, 600, 1800];

export interface QueueEntry {
  id: string;
  kind: string;
  method: string;
  path: string;
  body: Record<string, unknown> | null;
  queued_at: number;
  attempts: number;
  last_attempt_at: number | null;
  /** The ratchet counter this entry's MAC was computed at. Never re-issued. */
  counter: number;
}

export class QueueStore {
  constructor(readonly file: string) {
    const d = path.dirname(file);
    if (d) fs.mkdirSync(d, { recursive: true, mode: 0o700 });
    if (!fs.existsSync(file)) fs.closeSync(fs.openSync(file, 'w', 0o600));
  }

  enqueue(e: {
    kind: string;
    method: string;
    path: string;
    body?: Record<string, unknown> | null;
    counter: number;
  }): QueueEntry {
    const entry: QueueEntry = {
      id: crypto.randomBytes(16).toString('hex'),
      kind: e.kind,
      method: e.method,
      path: e.path,
      body: e.body ?? null,
      queued_at: Date.now() / 1000,
      attempts: 0,
      last_attempt_at: null,
      counter: e.counter,
    };
    // Append + fsync. An enqueue that is only in the page cache is not a
    // queue that "survives process death", which is the one property this
    // module exists for.
    const fd = fs.openSync(this.file, 'a', 0o600);
    try {
      fs.writeFileSync(fd, JSON.stringify(entry) + '\n');
      fs.fsyncSync(fd);
    } finally {
      fs.closeSync(fd);
    }
    return entry;
  }

  loadAll(): QueueEntry[] {
    let raw: string;
    try {
      raw = fs.readFileSync(this.file, 'utf8');
    } catch {
      return [];
    }
    const out: QueueEntry[] = [];
    for (const line of raw.split('\n')) {
      const t = line.trim();
      if (!t) continue;
      try {
        out.push(JSON.parse(t) as QueueEntry);
      } catch {
        // A torn last line from a crash mid-append. Skipped, as queue.py
        // skips a JSONDecodeError, rather than discarding the whole file.
        continue;
      }
    }
    return out;
  }

  replaceAll(entries: QueueEntry[]): void {
    const tmp = `${this.file}.tmp`;
    const fd = fs.openSync(tmp, 'w', 0o600);
    try {
      fs.writeFileSync(fd, entries.map((e) => JSON.stringify(e)).join('\n') + (entries.length ? '\n' : ''));
      fs.fsyncSync(fd);
    } finally {
      fs.closeSync(fd);
    }
    fs.renameSync(tmp, this.file);
  }

  count(): number {
    return this.loadAll().length;
  }
}

/** Seconds to wait after `attempts` failures. Saturates at the last step —
 *  a queue never gives up, because giving up is data loss dressed as
 *  cleanup. */
export function backoffSeconds(attempts: number): number {
  if (attempts <= 0) return 0;
  const i = Math.min(attempts, BACKOFF_SCHEDULE.length) - 1;
  return BACKOFF_SCHEDULE[i];
}

export function isDue(e: QueueEntry, nowMs = Date.now()): boolean {
  if (e.attempts === 0 || e.last_attempt_at === null) return true;
  return nowMs / 1000 - e.last_attempt_at >= backoffSeconds(e.attempts);
}
